### PR TITLE
Refactor collaborations into its own table

### PR DESCRIPTION
**Pull request by phsmit**
This PR introduces a new Collaboration struct which is a direct relation between a repository and a non-organization user. Introducing this table will make it much easier to refactor the access table.

### CODE REVIEW

File: models/models.go [hunk 1]
```diff
@@ -127,13 +127,65 @@ func SetEngine() (err error) {
...
+	return nil
+}
+
 func NewEngine() (err error) {
 	if err = SetEngine(); err != nil {
 		return err
 	}
+
+	var hasAccess, hasCollaboration, needMigration bool
+	if hasAccess, err = x.IsTableExist("access"); err != nil {
```
Comment (Unknwon) [inline models/models.go]: I would suppose `access` always exists(since initial Gogs version)?

Comment (phsmit) [inline models/models.go]: This is actually a disguised test to see if there is any tables at all. For a new installation no migration needs to be run

Comment (Unknwon) [inline models/models.go]: Right, but I think it costs nothing to `select` from empty `access` table, and we will have more organized code block(put `IsTableExist` and `migrateAccessToCollaboration ` together)?

Comment (phsmit) [inline models/models.go]: Good point. However, we can't put the two block together, as we want to test for the collaboration table before the Sync, and do the actual filling after the table has been synced.

Comment (Unknwon) [inline models/models.go]: Ah, yes. I'm stilling thinking on database version, so there is a function go through auto-migrate from the closest large version one by one until the current. It only requires only one person work on database struct change at one time, which I don't think is that hard?

Comment (phsmit) [inline models/models.go]: Ok. I'll try to make this and make a new PR. Let's see how that looks.

Comment (Unknwon) [inline models/models.go]: Sure, and it's completely fine to merge like this and let me do the version work.

**Comment by phsmit** [comment]
Superseded by #899

**Pull request closed without merging**